### PR TITLE
Improved LoGra with General two-stage compressor

### PR DESCRIPTION
**Pull request by sleepymalc**
## Description

1. Improving gradient hooks to fully avoid the gradient of parameter computation
2. Implement a general two-stage compressor (supporting LoGra and FactGraSS)

### 1. Motivation and Context

1. The current gradient hook still computes the gradient of the parameter, resulting in redundant computation.
2. The compressor is specific to LoGra, and can not integrate nicely with new compressor such as FactGraSS.

### 2. Summary of the change

1. Implement a monkey patch for a custom backward pass that avoids computing full parameter gradients.
2. Implement a general compressor class that handles two-stage compression.

### 3. What tests have been added/updated for the change?
- [x] N/A: No test will be added (please justify)
- [ ] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.
- [ ] Application test: If you wrote an example for the toolkit, this test should be added.
- [ ] Document test: If you added an external API, then you should check if the document is correctly generated.

### CODE REVIEW

File: dattri/algorithm/block_projected_if/__init__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"""LoGra utilities package.
```
Comment (jiaqima) [inline dattri/algorithm/block_projected_if/__init__.py]: LoGra -> BlockProjectedIF?

**Review by jiaqima — APPROVED** [review]
LGTM. Thanks!

File: dattri/algorithm/block_projected_if/__init__.py (hunk 1 above)
Comment (sleepymalc) [reply dattri/algorithm/block_projected_if/__init__.py]: Thanks for the catch.

File: test/dattri/algorithm/test_logra.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 from torch import nn
 from torch.utils.data import DataLoader, TensorDataset
 
-from dattri.algorithm.logra.logra import LoGraAttributor
+from dattri.algorithm import LoGraAttributor
```
[the remark this answers is not in the document]
Comment (sleepymalc) [reply test/dattri/algorithm/test_logra.py]: Sounds good

**Comment by sleepymalc** [comment]
I think it's ready, but I'll wait for @TheaperDeng to confirm.
